### PR TITLE
[SDK-574] Logs sorted out

### PR DESCRIPTION
**Pull request by turtledreams**
SDK-574

Logs have color now
Log level given at the start
Public func logs gives func info too
Public funcs are covered

### CODE REVIEW

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
Move the log call to a common "countly-common.js" module.

**Comment by ArtursKadikis** [comment]
this has merge conflicts

**Comment by turtledreams** [comment]
image

Module has been created and everything works same.

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
"countly-common" needs to be reworked to an object. 
The "require" path should be shortened.

Create a callback for "isLoggingEnabled"

File: lib/countly-bulk-user.js [hunk 1]
```diff
@@ -61,12 +61,12 @@ function CountlyBulkUser(conf) {
     cc.debugBulkUser = conf.debug;
 
     if (!conf.device_id) {
-        cc.log("device_id is missing");
+        cc.log(cc.logLevelEnums.WARNING, "device_id is missing");
```
Comment (ArtursKadikis) [inline lib/countly-bulk-user.js:64]: should be "error"

File: lib/countly-bulk-user.js [hunk 2]
```diff
@@ -61,12 +61,12 @@ function CountlyBulkUser(conf) {
...
         return;
     }
 
     if (!conf.server) {
-        cc.log("server instance is missing");
+        cc.log(cc.logLevelEnums.WARNING, "server instance is missing");
```
Comment (ArtursKadikis) [inline lib/countly-bulk-user.js:69]: should be "error"

File: lib/countly-bulk-user.js [hunk 3]
```diff
@@ -425,11 +425,11 @@ function CountlyBulkUser(conf) {
     this.report_feedback = function(feedback, timestamp) {
         if (this.check_consent("star-rating") || this.check_consent("feedback")) {
             if (!feedback.widget_id) {
-                cc.log("Feedback must contain widget_id property");
+                cc.log(cc.logLevelEnums.WARNING, "report_feedback, Feedback must contain widget_id property");
```
Comment (ArtursKadikis) [inline lib/countly-bulk-user.js:428]: should be "error"

File: lib/countly-bulk-user.js [hunk 4]
```diff
@@ -425,11 +425,11 @@ function CountlyBulkUser(conf) {
...
                 return;
             }
             if (!feedback.rating) {
-                cc.log("Feedback must contain rating property");
+                cc.log(cc.logLevelEnums.WARNING, "report_feedback, Feedback must contain rating property");
```
Comment (ArtursKadikis) [inline lib/countly-bulk-user.js:432]: should be "error"

File: lib/countly-bulk-user.js [hunk 5]
```diff
@@ -466,7 +466,7 @@ function CountlyBulkUser(conf) {
             var props = ["type", "name", "stz", "etz", "apm_metrics", "apm_attr"];
             for (var i = 0; i < props.length; i++) {
                 if (props[i] !== "apm_attr" && typeof trace[props[i]] === "undefined") {
-                    cc.log("APM trace must have a", props[i]);
+                    cc.log(cc.logLevelEnums.WARNING, "report_trace, APM trace must have a", props[i]);
```
Comment (ArtursKadikis) [inline lib/countly-bulk-user.js:469]: should be "error"

File: lib/countly-bulk.js [hunk 6]
```diff
@@ -67,12 +68,12 @@ function CountlyBulk(conf) {
 
     cc.debugBulk = conf.debug || false;
     if (!conf.app_key) {
-        cc.log("app_key is missing");
+        cc.log(cc.logLevelEnums.WARNING, "app_key is missing");
```
Comment (ArtursKadikis) [inline lib/countly-bulk.js:71]: should be "error"

File: lib/countly-bulk.js [hunk 7]
```diff
@@ -67,12 +68,12 @@ function CountlyBulk(conf) {
...
         return;
     }
 
     if (!conf.url) {
-        cc.log("url is missing");
+        cc.log(cc.logLevelEnums.WARNING, "url is missing");
```
Comment (ArtursKadikis) [inline lib/countly-bulk.js:76]: should be "error"

File: lib/countly-bulk.js [hunk 8]
```diff
@@ -120,7 +121,7 @@ function CountlyBulk(conf) {
     this.add_request = function(query) {
         if (cluster.isMaster) {
             if (!query.device_id) {
-                cc.log("device_id is missing", query);
+                cc.log(cc.logLevelEnums.WARNING, "add_request, device_id is missing", query);
```
Comment (ArtursKadikis) [inline lib/countly-bulk.js:124]: should be "error"

File: lib/countly-bulk.js [hunk 9]
```diff
@@ -129,7 +130,7 @@ function CountlyBulk(conf) {
             }
 
             if ((query.timestamp + "").length !== 13 && (query.timestamp + "").length !== 10) {
-                cc.log("incorrect timestamp format", query);
+                cc.log(cc.logLevelEnums.WARNING, "add_request, incorrect timestamp format", query);
```
Comment (ArtursKadikis) [inline lib/countly-bulk.js:133]: should be "error"

File: lib/countly-bulk.js [hunk 10]
```diff
@@ -166,7 +167,7 @@ function CountlyBulk(conf) {
             for (var i in requests) {
                 query = requests[i];
                 if (!query.device_id) {
-                    cc.log("device_id is missing", query);
+                    cc.log(cc.logLevelEnums.WARNING, "add_bulk_request, device_id is missing", query);
```
Comment (ArtursKadikis) [inline lib/countly-bulk.js:170]: should be "error"

File: lib/countly-bulk.js [hunk 11]
```diff
@@ -175,7 +176,7 @@ function CountlyBulk(conf) {
                 }
 
                 if ((query.timestamp + "").length !== 13 && (query.timestamp + "").length !== 10) {
-                    cc.log("incorrect timestamp format", query);
+                    cc.log(cc.logLevelEnums.WARNING, "add_bulk_request, incorrect timestamp format", query);
```
Comment (ArtursKadikis) [inline lib/countly-bulk.js:179]: should be "error"

File: lib/countly-bulk.js [hunk 12]
```diff
@@ -216,12 +217,12 @@ function CountlyBulk(conf) {
     **/
     this.add_event = function(device_id, event) {
         if (!device_id) {
-            cc.log("device_id is missing");
+            cc.log(cc.logLevelEnums.WARNING, "add_event, device_id is missing");
```
Comment (ArtursKadikis) [inline lib/countly-bulk.js:220]: should be "error"

File: lib/countly-bulk.js [hunk 13]
```diff
@@ -216,12 +217,12 @@ function CountlyBulk(conf) {
...
             return;
         }
 
         if (!event.key) {
-            cc.log("Event must have key property");
+            cc.log(cc.logLevelEnums.WARNING, "add_event, Event must have key property");
```
Comment (ArtursKadikis) [inline lib/countly-bulk.js:225]: should be "error"

File: lib/countly-common.js [hunk 14]
```diff
@@ -75,17 +75,43 @@ var cc = {
...
+                color = '\x1b[34m%s\x1b[0m';
+            }
+            else {
+                //default log level is info
+                level = this.logLevelEnums.INFO;
+            }
+
             // eslint-disable-next-line no-console
-            console.log(Array.prototype.slice.call(arguments).join("\n"));
+            console.log(color, level + message, Array.prototype.slice.call(args).join("\n"));
```
Comment (ArtursKadikis) [inline lib/countly-common.js:111]: depending on the log level, you should use the available "logWarning" and "logError" functions

File: lib/countly.js [hunk 15]
```diff
@@ -169,10 +169,10 @@ Countly.Bulk = Bulk;
             }
 
             if (Countly.url === "") {
-                cc.log("Please provide server URL");
+                cc.log(cc.logLevelEnums.WARNING, "Please provide server URL");
```
Comment (ArtursKadikis) [inline lib/countly.js:172]: should be "error"

File: lib/countly.js [hunk 16]
```diff
@@ -217,23 +217,24 @@ Countly.Bulk = Bulk;
...
                 }
                 else {
-                    cc.log("Feature name " + i + " is already reserved");
+                    cc.log(cc.logLevelEnums.WARNING, "group_features, Feature name " + i + " is already reserved");
                 }
             }
         }
         else {
-            cc.log("Incorrect features: " + features);
+            cc.log(cc.logLevelEnums.WARNING, "group_features, Incorrect features: " + features);
```
Comment (ArtursKadikis) [inline lib/countly.js:237]: should be "error"

File: lib/countly.js [hunk 17]
```diff
@@ -251,7 +252,7 @@ Countly.Bulk = Bulk;
             return (consents[feature] && consents[feature].optin) ? true : false;
         }
         else {
-            cc.log("No feature available for " + feature);
+            cc.log(cc.logLevelEnums.WARNING, "check_consent, No feature available for " + feature);
```
Comment (ArtursKadikis) [inline lib/countly.js:255]: should be "error"

File: lib/countly.js [hunk 18]
```diff
@@ -517,7 +524,7 @@ Countly.Bulk = Bulk;
     */
     function add_cly_events(event) {
         if (!event.key) {
-            cc.log("Event must have key property");
+            cc.log(cc.logLevelEnums.WARNING, "Event must have key property");
```
Comment (ArtursKadikis) [inline lib/countly.js:527]: should be "error"

File: lib/countly.js [hunk 19]
```diff
@@ -561,13 +570,15 @@ Countly.Bulk = Bulk;
             event = {key: event};
         }
         if (!event.key) {
-            cc.log("Event must have key property");
+            cc.log(cc.logLevelEnums.WARNING, "end_event, Event must have key property");
```
Comment (ArtursKadikis) [inline lib/countly.js:573]: should be "error"

File: lib/countly.js [hunk 20]
```diff
@@ -745,12 +756,16 @@ Countly.Bulk = Bulk;
...
+
             }
             else if (campaign_id) {
                 toRequestQueue({campaign_id: campaign_id});
+                cc.log(cc.logLevelEnums.INFO, "report_conversion, Conversion reported");
+
             }
             else {
-                cc.log("No campaign data found");
+                cc.log(cc.logLevelEnums.WARNING, "report_conversion, No campaign data found");
```
Comment (ArtursKadikis) [inline lib/countly.js:768]: should be "error"

File: lib/countly.js [hunk 21]
```diff
@@ -769,11 +784,11 @@ Countly.Bulk = Bulk;
     Countly.report_feedback = function(feedback) {
         if (Countly.check_consent("star-rating") || Countly.check_consent("feedback")) {
             if (!feedback.widget_id) {
-                cc.log("Feedback must contain widget_id property");
+                cc.log(cc.logLevelEnums.WARNING, "report_feedback, Feedback must contain widget_id property");
```
Comment (ArtursKadikis) [inline lib/countly.js:787]: should be "error"

File: lib/countly.js [hunk 22]
```diff
@@ -769,11 +784,11 @@ Countly.Bulk = Bulk;
...
                 return;
             }
             if (!feedback.rating) {
-                cc.log("Feedback must contain rating property");
+                cc.log(cc.logLevelEnums.WARNING, "report_feedback, Feedback must contain rating property");
```
Comment (ArtursKadikis) [inline lib/countly.js:791]: should be "error"

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
Some lines have "trailing spaces" those should be fixed.

File: lib/countly.js [hunk 23]
```diff
@@ -169,10 +169,10 @@ Countly.Bulk = Bulk;
             }
 
             if (Countly.url === "") {
-                cc.log("Please provide server URL");
+                cc.log(cc.logLevelEnums.ERROR, "Please provide server URL");
             }
             else {
-                cc.log("Countly initialized");
+                cc.log("info", "Countly initialized");
```
Comment (ArtursKadikis) [inline lib/countly.js:175]: the "level" tag is hardcoded and doesn't reference the variable

File: lib/countly-common.js [hunk 24]
```diff
@@ -75,17 +75,47 @@ var cc = {
...
+                if (level === this.logLevelEnums.DEBUG) {
+                    //cyan
+                    color = '\x1b[36m%s\x1b[0m';
+                }
+                else if (level === this.logLevelEnums.VERBOSE) {
+                    //blue
+                    color = '\x1b[34m%s\x1b[0m';
+                }
+                else {
+                    //default log level is info
```
Comment (ArtursKadikis) [inline lib/countly-common.js:110]: make "debug" the default

File: lib/countly.js [hunk 25]
```diff
@@ -399,7 +406,7 @@ Countly.Bulk = Bulk;
     Countly.begin_session = function(noHeartBeat) {
         if (Countly.check_consent("sessions")) {
             if (!sessionStarted) {
-                cc.log("Session started");
+                cc.log("info", "begin_session, Session started");
```
Comment (ArtursKadikis) [inline lib/countly.js:409]: log level is not linking to the enum

File: lib/countly.js [hunk 26]
```diff
@@ -286,7 +293,7 @@ Countly.Bulk = Bulk;
     * @param {string|array} feature - name of the feature, possible values, "sessions","events","views","crashes","attribution","users" or customly provided through {@link Countly.group_features}
     */
     Countly.add_consent = function(feature) {
-        cc.log("Adding consent for " + feature);
+        cc.log("info", "add_consent, Adding consent for " + feature);
```
Comment (ArtursKadikis) [inline lib/countly.js:296]: log level is not linking to the enum

File: lib/countly.js [hunk 27]
```diff
@@ -333,7 +340,7 @@ Countly.Bulk = Bulk;
     * @param {string|array} feature - name of the feature, possible values, "sessions","events","views","crashes","attribution","users" or custom provided through {@link Countly.group_features}
     */
     Countly.remove_consent = function(feature) {
-        cc.log("Removing consent for " + feature);
+        cc.log("info", "remove_consent, Removing consent for " + feature);
```
Comment (ArtursKadikis) [inline lib/countly.js:343]: log level is not linking to the enum

File: lib/countly.js [hunk 28]
```diff
@@ -421,7 +428,7 @@ Countly.Bulk = Bulk;
     Countly.session_duration = function(sec) {
         if (Countly.check_consent("sessions")) {
             if (sessionStarted) {
-                cc.log("Session extended", sec);
+                cc.log("info", "session_duration, Session extended", sec);
```
Comment (ArtursKadikis) [inline lib/countly.js:431]: log level is not linking to the enum

File: lib/countly.js [hunk 29]
```diff
@@ -468,7 +475,7 @@ Countly.Bulk = Bulk;
                 var oldId = Countly.device_id;
                 Countly.device_id = newId;
                 storeSet("cly_id", Countly.device_id);
-                cc.log("Changing id");
+                cc.log("info", "change_id, Changing ID");
```
Comment (ArtursKadikis) [inline lib/countly.js:478]: log level is not linking to the enum

File: lib/countly.js [hunk 30]
```diff
@@ -1306,7 +1346,7 @@ Countly.Bulk = Bulk;
     */
     function makeRequest(url, api, params, callback) {
         try {
-            cc.log("Sending HTTP request");
+            cc.log('info', "Sending HTTP request");
```
Comment (ArtursKadikis) [inline lib/countly.js:1349]: log level is not linking to the enum

File: lib/countly.js [hunk 31]
```diff
@@ -435,7 +442,7 @@ Countly.Bulk = Bulk;
         if (Countly.check_consent("sessions")) {
             if (sessionStarted) {
                 sec = sec || cc.getTimestamp() - lastBeat;
-                cc.log("Ending session");
+                cc.log("info", "end_session, Ending session");
```
Comment (ArtursKadikis) [inline lib/countly.js:445]: log level is not linking to the enum

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
Requires a few tweaks to choosing colors

**Review by ArtursKadikis — APPROVED** [review]

**Pull request merged**